tests/unit/steps/test_steps_plan.py

```diff
@@ -701,3 +689,89 @@ async def test_get_upgradable_hypervisors_machines(
     assert {
         hypervisor.machine_id for hypervisor in hypervisors_possible_to_upgrade
     } == expected_result
+
+
+@pytest.mark.parametrize("cli_backup", [True, False])
+def test_generate_common_plan(cli_backup, cli_args, model):
+    cli_args.backup = cli_backup
+    mock_analysis_result = MagicMock(spec=Analysis)()
+    mock_analysis_result.current_cloud_os_release = OpenStackRelease("ussuri")
+    mock_analysis_result.model = model
+
+    target = OpenStackRelease("victoria")
+
+    expected_plan = UpgradePlan("Upgrade cloud from 'ussuri' to 'victoria'")
+    expected_plan.add_step(
+        PreUpgradeStep(
+            description="Verify that all OpenStack applications are in idle state",
+            parallel=False,
+            coro=mock_analysis_result.model.wait_for_active_idle(
+                timeout=11, idle_period=10, raise_on_blocked=True
+            ),
+        )
+    )
+
+    common_plan = cou_plan.generate_common_plan(target, mock_analysis_result, cli_args)
+
+    if cli_backup:
+        expected_plan.add_step(
+            PreUpgradeStep(
+                description="Backup mysql databases",
+                parallel=False,
+                coro=backup(model),
+            )
+        )
+
+    assert common_plan == expected_plan
+
+
+@patch("cou.steps.plan.create_upgrade_group")
+def test_generate_control_plane_plan(mock_create_upgrade_group):
+    target = OpenStackRelease("victoria")
+    keystone = MagicMock(spec_set=OpenStackApplication)()
+
+    cou_plan._generate_control_plane_plan(target, [keystone], False)
+
+    assert mock_create_upgrade_group.call_count == 2
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "upgrade_group",
+    [(CONTROL_PLANE, DATA_PLANE, HYPERVISORS, None)],
+)
+@patch("cou.steps.plan.generate_common_plan")
+@patch("cou.steps.plan.determine_upgrade_target")
+@patch("cou.steps.plan.pre_plan_sanity_checks")
+@patch("cou.steps.plan._generate_data_plane_plan")
+@patch("cou.steps.plan._generate_control_plane_plan")
+async def test_generate_plan_control_and_data_plane(
+    mock_control_plane,
+    mock_data_plane,
+    mock_pre_plan_sanity_checks,
+    mock_determine_upgrade_target,
+    mock_generate_common_plan,
+    cli_args,
+    upgrade_group,
+):
+    cli_args.upgrade_group = upgrade_group
+
+    mock_analysis_result = MagicMock(spec=Analysis)()
+
+    await cou_plan.generate_plan(mock_analysis_result, cli_args)
+
+    mock_pre_plan_sanity_checks.assert_called_once()
+    mock_determine_upgrade_target.assert_called_once()
+    mock_generate_common_plan.assert_called_once()
+
+    if upgrade_group is None:
+        mock_control_plane.assert_called_once()
+        mock_data_plane.assert_called_once()
+
+    if upgrade_group == CONTROL_PLANE:
+        mock_control_plane.assert_called_once()
+        mock_data_plane.assert_not_called()
+
+    if upgrade_group in {DATA_PLANE, HYPERVISORS}:
+        mock_control_plane.assert_not_called()
+        mock_data_plane.assert_called_once()
```

Comment:
non-blocking: I'm not a fan of parameterize with conditional checks. They are better separated into different test functions.